seismic/ASDFdatabase/asdf_preprocess.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/env python
+"""
+Description:
+    Reads waveforms from an ASDF file, optionally applies instrument response correction,
+    resamples and outputs them to another ASDF file. This preprocessing is crucial for
+    large-scale studies involving > 10000 Green's Functions, e.g. in ambient noise
+    tomography. This approach significantly reduces IO bottlenecks and computational
+    costs associated with having to apply instrument response corrections on data from
+    a given station in alternative workflows.
+References:
+
+CreationDate:   18/07/19
+Developer:      rakib.hassan@ga.gov.au
+
+Revision History:
+    LastUpdate:     18/07/19   RH
+    LastUpdate:     dd/mm/yyyy  Who     Optional description
+"""
+
+import click
+import os
+from mpi4py import MPI
+import pyasdf
+from tqdm import tqdm
+from obspy import UTCDateTime, read_inventory, Inventory, Stream
+from collections import defaultdict
+from obspy.core.util.misc import get_window_times
+import gc
+from obspy.core.util.misc import limit_numpy_fft_cache
+
+def split_list(lst, npartitions):
+    k, m = divmod(len(lst), npartitions)
+    return [lst[i * k + min(i, m):(i + 1) * k + min(i + 1, m)] for i in range(npartitions)]
+# end func
+
+def getStationInventory(master_inventory, inventory_cache, netsta):
+    netstaInv = None
+    if (master_inventory):
+        if (inventory_cache is None): inventory_cache = defaultdict(list)
+        net, sta = netsta.split('.')
+
+        if (isinstance(inventory_cache[netsta], Inventory)):
+            netstaInv = inventory_cache[netsta]
+        else:
+            inv = master_inventory.select(network=net, station=sta)
+            if(len(inv.networks)):
+                inventory_cache[netsta] = inv
+                netstaInv = inv
+            # end if
+        # end if
+    # end if
+
+    return netstaInv, inventory_cache
+# end func
+
+def create_station_asdf(input_asdf, output_folder, resample_rate,
+                        instrument_response_inventory, instrument_response_output, water_level):
+    # mpi attributes
+    comm = MPI.COMM_WORLD
+    nproc = comm.Get_size()
+    rank = comm.Get_rank()
+
+    # input asdf file
+    ids = pyasdf.ASDFDataSet(input_asdf, mode='r')
+
+    # get stations
+    stations = ids.get_all_coordinates().keys()
+
+    # local work-load
+    stations = split_list(stations, nproc)[rank]
+
+    # read inventory
+    stationInvCache = None
+    # read inventory
+    inv = None
+    try:
+        inv = read_inventory(instrument_response_inventory)
+    except Exception as e:
+        print (e)
+        raise RuntimeError('Failed to read inventory: %s' % (instrument_response_inventory))
+    # end try
+
+    for s in stations:
```

Comment:
Please add a comment documenting why garbage collection is needed in this loop. In general it's something one shouldn't have to worry about.